src/NativeScript/Marshalling/Reference/ReferenceConstructor.cpp

```diff
@@ -27,32 +27,32 @@ static EncodedJSValue JSC_HOST_CALL constructReference(ExecState* execState) {
     GlobalObject* globalObject = jsCast<GlobalObject*>(execState->lexicalGlobalObject());
     ReferenceInstance* result;
 
-    if (execState->argumentCount() == 2) {
-        JSValue type = execState->uncheckedArgument(0);
-        JSValue value = execState->uncheckedArgument(1);
+    JSValue maybeType = execState->argument(0);
+    const FFITypeMethodTable* ffiTypeMethodTable;
+    if (tryGetFFITypeMethodTable(maybeType, &ffiTypeMethodTable)) {
+        void* handle = nullptr;
+        bool adopted = true;
 
-        const FFITypeMethodTable* ffiTypeMethodTable;
-        if (!tryGetFFITypeMethodTable(type, &ffiTypeMethodTable)) {
-            return throwVMError(execState, createError(execState, WTF::ASCIILiteral("Not a valid type object is passed as parameter.")));
-        }
-
-        bool hasHandle;
-        void* handle = tryHandleofValue(value, &hasHandle);
-        PointerInstance* pointer;
-        if (hasHandle) {
-            pointer = jsCast<PointerInstance*>(globalObject->interop()->pointerInstanceForPointer(execState->vm(), handle));
+        if (execState->argumentCount() == 2) {
+            JSValue value = execState->uncheckedArgument(1);
+            if (PointerInstance* pointer = jsDynamicCast<PointerInstance*>(value)) {
```

Comment:
So, `RecordInstance` and `ReferenceInstance` can be special-handled like `PointerInstance` is now?